src/pystack/_pystack/mem.cpp

```diff
@@ -344,42 +344,99 @@ CorefileRemoteMemoryManager::CorefileRemoteMemoryManager(
 {
     CoreFileExtractor extractor{d_analyzer};
     d_shared_libs = extractor.ModuleInformation();
+
+    const char* filename = d_analyzer->d_filename.c_str();
+    int fd = open(filename, O_RDONLY);
+
+    StatusCode ret = readCorefile(fd, filename);
+    close(fd);
+
+    if (ret == StatusCode::ERROR) {
+        throw RemoteMemCopyError();
+    }
+}
+
+CorefileRemoteMemoryManager::~CorefileRemoteMemoryManager()
+{
+    if (munmap(corefile_data, corefile_size) == -1) {
+        LOG(ERROR) << "Failed to un-mmap a file " << d_analyzer->d_filename.c_str();
+    }
+}
+
+CorefileRemoteMemoryManager::StatusCode
+CorefileRemoteMemoryManager::readCorefile(int fd, const char* filename)
+{
+    if (fd == -1) {
+        LOG(ERROR) << "Failed to open a file " << filename;
+        return StatusCode::ERROR;
+    }
+
+    struct stat fileInfo = {0};
+
+    if (fstat(fd, &fileInfo) == -1) {
+        LOG(ERROR) << "Failed to get a file size for a file " << filename;
+        return StatusCode::ERROR;
+    }
+
+    if (fileInfo.st_size == 0) {
+        LOG(ERROR) << "File " << filename << " is empty";
+        return StatusCode::ERROR;
+    }
+
+    corefile_size = fileInfo.st_size;
+    corefile_data = reinterpret_cast<char*>(mmap(0, corefile_size, PROT_READ, MAP_PRIVATE, fd, 0));
+
+    if (corefile_data == MAP_FAILED) {
+        LOG(ERROR) << "Failed to mmap a file " << filename;
+        return StatusCode::ERROR;
+    }
+
+    int madvise_result = madvise(corefile_data, corefile_size, MADV_RANDOM);
+
+    if (madvise_result == -1) {
+        LOG(ERROR) << "Madvise for a file " << filename << " failed";
+        return StatusCode::ERROR;
+    }
+
+    return StatusCode::SUCCESS;
 }
 
 ssize_t
 CorefileRemoteMemoryManager::copyMemoryFromProcess(remote_addr_t addr, size_t size, void* destination)
         const
 {
-    const std::string* filename = nullptr;
     off_t offset_in_file = 0;
 
-    StatusCode ret = getMemoryLocationFromCore(addr, &filename, &offset_in_file);
+    StatusCode ret = getMemoryLocationFromCore(addr, &offset_in_file);
 
-    if (ret == StatusCode::ERROR) {
-        // The memory may be in the data segment of some shared library
-        getMemoryLocationFromElf(addr, &filename, &offset_in_file);
+    if (ret == StatusCode::SUCCESS) {
```

Comment:
Let's add an assert here that offset_in_file is smaller than the mmaped chunk

Reply:
Should it be an assert, or should we `throw InvalidRemoteAddress()`? I'm not sure if that's an invariant or something that depends on the contents of the core file...